### PR TITLE
Adds Sabrage

### PR DESCRIPTION
**Pull request by Sealed101**
## About The Pull Request
This PR adds an interaction to champagne bottles that allows the user to pull off the cork with an edged (for code purposes, sharp) item. The sharper the item, the better your chance is to pull the stunt off. Minimum force is 5, which lets you do it with a captain's pen, for instance. It's slower than just removing the cork by hand, as to simulate the player removing the foil before taking the swing. Heads of staff get a +20% chance to succeed due to being officers in some capacity. A skillchip that can be bought from the Library Play Room vendor increases the chance by +35%. Failing the stunt will smash the bottle in your hands, spilling the contents all over you. This will produce a bottle with the neck broken off, instead of its bottom.

Pulling the stunt off provides a weak mood buff for the user, and a weaker mood buff for any witnesses. However, if someone else pulls the sabrage off after you, you will get your success buff removed. Think of it like pulling any other stunt and challenging someone else to do it, only for them to succeed as well. Makes you bitter, doesn't it? Failing the stunt also produces a weak mood debuff to the user.

image

Captains now have a chance to receive a "cursed" champagne bottle in the mail, that will not take the blade's force into account when trying to sabrage.

### DLC: frothing!
Champagne bottles now can froth when uncorked!
Froth comes in three intensity flavours, causing the bottle to let loose up to 15u of its contents.

For the usual in-hand uncorking, it's just a 50/50 chance to have some booze froth.

For the sabrage, the less chance you had to succeed in pulling the stunt off, the more powerful the froth will be. If you were guaranteed a sabrage by having 100%+ chance on your sabrage rolls, it won't froth at all!
Here's how it works with sabrage
<details>
<summary>click for video (GACHI BGM WARNING)</summary>

https://user-images.githubusercontent.com/75863639/198887676-e0123eae-c3ec-4377-a559-8b9df751d96f.mp4

</details>

here's how it works with the usual in-hand use method (i can't record the in-hand panel in good quality so the bottles are on the floor but you get the idea)

<details>
<summary>click for video (GACHI BGM WARNING)</summary>

https://user-images.githubusercontent.com/75863639/198887715-0477c6f1-4be9-4f42-ac87-f898926e6f92.mp4

</details>

## Why It's Good For The Game
image

## Changelog
:cl:
add: You can now perform an overly-complicated cork removal ritual known as Sabrage. Swing at a champagne bottle with a sharp enough blade to strike the cork off the bottle!
add: Added a Sabrage skill chip to the Library Play Room Vendor.
add: Heads of staff are more likely to pull this stunt off.
/:cl:

### CODE REVIEW

**Comment by Sealed101** [comment]
there is no check for the bottle to be in-hand to pull off the sabrage, so in theory one can assemble an array of champagne bottles and shoot out a barrage of corks at their enemies. sounds funny but not quite true to the source material

**Review by MrMelbert — COMMENTED** [review]
I've been wanting to add this for ages

File: code/modules/library/skill_learning/skillchip.dm [hunk 1]
```diff
@@ -450,6 +450,16 @@
 	activate_message = span_notice("You feel that you can recognize special, minute details on ID cards.")
 	deactivate_message = span_notice("Was there something special about certain IDs?")
 
+/obj/item/skillchip/sabrage
```
[the remark this answers is not in the document]
Comment (Sealed101) [reply code/modules/library/skill_learning/skillchip.dm:453]: captains already have a +20% buff, and their officer sabre is a 15 force item, resulting in 75 + 20 chance. even with a pen it's somewhat reliable

**Comment by Likteer** [comment]
There is really no point in making it all that complicated, no skillchips, no different petty mood shifts, no universal action for sharp enough objects.

I think you're too self-conscious about making it balanced enough, Most people playing this game don't even know about the **regular** cork launch.

**Comment by optimumtact** [comment]
> I believe I have a better idea.

I believe you should shut up

**Comment by Sealed101** [comment]
> _a different idea on implementing this_

i like mine better

**Comment by LarissaMayrink** [comment]
> I believe I have a better idea.
> 
> * Only sabre works on the bottle.

captain pen sabrage best sabrage for maximum style, more skillchips is also pretty nice given we don't have that many just /fun/ ones yet

**Comment by Likteer** [comment]
> i like mine better

Understandable, have a good day

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 2]
```diff
@@ -473,30 +478,80 @@
...
+
+/obj/item/reagent_containers/cup/glass/bottle/champagne/attackby(obj/item/attacking_item, mob/living/user, params)
+	. = ..()
+	if(spillable)
+		return
+	if(attacking_item.get_sharpness())
+		if(attacking_item.force < 5)
+			balloon_alert(user, "not strong enough!")
+			return
+		else
```
Comment (tralezab) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:504]: you return early, but STILL put in an else case, get rid of the else and unindent the following block

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 3]
```diff
@@ -473,30 +478,80 @@
...
+			span_nicegreen("You elegantly slice the cork off of [src], causing it to fly off the bottle with great force."))
+		for(var/mob/living/carbon/stunt_witness in view(7, user))
+			for(var/find_previous_showoff in stunt_witness.mob_mood.mood_events) //if someone who did a sabrage before us saw us doing it, clear their success mood event
+				if(find_previous_showoff == "sabrage_success")
+					stunt_witness.clear_mood_event(find_previous_showoff)
+					continue
+			if(stunt_witness == user)
+				stunt_witness.add_mood_event("sabrage_success", /datum/mood_event/sabrage_success)
+				continue
+			else
```
Comment (tralezab) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:543]: if it continues it doesn't execute the following code, you don't need an else case here

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 4]
```diff
@@ -473,30 +478,80 @@
...
+		else
+			icon_state = "[base_icon_state]_popped"
 	else
 		icon_state = base_icon_state
 
-/obj/item/reagent_containers/cup/glass/bottle/champagne/proc/pop_cork(mob/user)
-	user.visible_message(span_danger("[user] loosens the cork of [src] causing it to pop out of the bottle with great force."), \
-		span_nicegreen("You elegantly loosen the cork of [src] causing it to pop out of the bottle with great force."))
+/obj/item/reagent_containers/cup/glass/bottle/champagne/proc/pop_cork(mob/living/user, sabrage)
+	if(sabrage)
```
Comment (tralezab) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:531]: switch this and early return! if(!sabrage) throw the message and return

**Review by tralezab — CHANGES_REQUESTED** [review]
gotta roast you a little on the early returns, ping me on discord if you're confused

**Review by Twaticus — APPROVED** [review]

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]
I feel like the container should lose a little bit of its reagents (maybe with a random chance) when being successful sabraged. (i've never seen this done without the bottle losing some of its liquid)

**Comment by Sealed101** [comment]
i think tying it to having or not having the skillchip should be fine. i will also try to make it froth when losing the contents, cuz that would be rad. we'll see

**Comment by Sealed101** [comment]
frothing is in!
tl;dr the less chance to succeed you had, the higher the froth's intensity.
here's how it works with sabrage
<details>
<summary>click for video (GACHI BGM WARNING)</summary>

https://user-images.githubusercontent.com/75863639/198887676-e0123eae-c3ec-4377-a559-8b9df751d96f.mp4

</details>

here's how it works with the usual in-hand use method (i can't record the in-hand panel in good quality so the bottles are on the floor but you get the idea)

<details>
<summary>click for video (GACHI BGM WARNING)</summary>

https://user-images.githubusercontent.com/75863639/198887715-0477c6f1-4be9-4f42-ac87-f898926e6f92.mp4

</details>

**Review by Imaginos16 — APPROVED** [review]
Yea

**Comment by LarissaMayrink** [comment]
oh no he used the cum champagne animations

https://media.discordapp.net/attachments/326849939240124419/1036239134265716736/froth_white3.gif

**Comment by Sealed101** [comment]
> oh no he used the cum champagne animations
> 
> https://media.discordapp.net/attachments/326849939240124419/1036239134265716736/froth_white3.gif

those are the whiter ones, i toned it down for the pr.

if you see ejaculate in champagne froth, this is on you

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 5]
```diff
@@ -473,30 +510,100 @@
...
+		return pop_cork(user, sabrage = FALSE, froth_severity = pick(0, 1))
+
+/obj/item/reagent_containers/cup/glass/bottle/champagne/attackby(obj/item/attacking_item, mob/living/user, params)
+	. = ..()
+
+	if(spillable)
+		return
+
+	if(!attacking_item.get_sharpness())
+		return
```
Comment (Sealed101) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:535]: might have to make some kind of blacklist for some of the more stupid items that you now can use for this. breaking a bottle and then sabraging with it is funny but stupid imo

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 6]
```diff
@@ -473,30 +510,100 @@
...
+
+	if(attacking_item.force < 5)
+		balloon_alert(user, "not strong enough!")
+		return
```
Comment (Sealed101) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:539]: this may also be excessive, as something like a syringe is also sharp and will trigger this alert while also saying that you can't draw/add reagents to the bottle, which might lead to some misinterpretation

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 7]
```diff
@@ -473,30 +510,100 @@
...
+				if(34 to 66)
+					severity_to_pass = 2
+				if(67 to 99)
+					severity_to_pass = 1
+		return pop_cork(user, sabrage = TRUE, froth_severity = severity_to_pass)
+	else //you dun goofed
+		user.visible_message(span_danger("[user] fumbles the sabrage and cuts [src] in half, spilling it over themselves!"), \
+			span_danger("You fail your stunt and cut [src] in half, spilling it over you!"))
+		user.add_mood_event("sabrage_fail", /datum/mood_event/sabrage_fail)
+		return smash(user, user, ranged = FALSE, break_top = TRUE)
```
Comment (Sealed101) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:569]: the problem with this is that failing sabrage while the bottle is not in your hand will put the broken bottle in your hand. i also did not test any of this with TK (agony)

**Review by Sealed101 — COMMENTED** [review]
thinking out loud

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm (hunk 7 above)
Comment (Sealed101) [reply code/modules/reagents/reagent_containers/cups/glassbottle.dm:569]: TK findings:
1. Failing does teleport the bottle into the user's hand. big sad
2. The cork cannot hit the user unless it gets a ricochet off something. bigger sad

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/reagents/reagent_containers/cups/glassbottle.dm [hunk 8]
```diff
@@ -137,10 +173,14 @@
 
 /// Mimics the appearance and properties of the passed in bottle.
 /// Takes the broken bottle to mimic, and the thing the bottle was broken agaisnt as args
-/obj/item/broken_bottle/proc/mimic_broken(obj/item/reagent_containers/cup/glass/to_mimic, atom/target)
+/obj/item/broken_bottle/proc/mimic_broken(obj/item/reagent_containers/cup/glass/to_mimic, atom/target, break_top)
 	icon_state = to_mimic.icon_state
 	var/icon/drink_icon = new(to_mimic.icon, icon_state)
-	drink_icon.Blend(broken_outline, ICON_OVERLAY, rand(5), 1)
+	if(break_top) //if the bottle breaks it's top off instead of the bottom
```
Comment (Sealed101) [inline code/modules/reagents/reagent_containers/cups/glassbottle.dm:179]:
```suggestion
	if(break_top) //if the bottle breaks its top off instead of the bottom
```

**Review by MrMelbert — APPROVED** [review]
